File: structural/C++/Decorator.cpp
#include <iostream>
#include <string>

class Widget
{
public:
	virtual void draw() = 0;
	virtual std::string description() = 0;
};

class Window : public Widget
{
public:
	void draw() override
	{
		std::cout << "Drawing the Window" << std::endl;
	}

	std::string description() override
	{
		return "Window";
	}
};

class WindowDecorator : public Window
{
public:
	WindowDecorator(Widget* windowToBeDecorated)
		: m_WindowToBeDecorated(windowToBeDecorated)
	{
	}

	void draw()
	{
		m_WindowToBeDecorated->draw(); // Delegation
	}

	std::string description()
	{
		return m_WindowToBeDecorated->description(); // Delegation
	}

private:
	Widget* m_WindowToBeDecorated;
};

class VerticalScrollBarDecorator : public WindowDecorator
{
public:
	VerticalScrollBarDecorator(Widget* windowToDecorate)
		: WindowDecorator(windowToDecorate)
	{
	}

	void draw() override
	{
		WindowDecorator::draw();
		drawVerticalScrollBar();
	}

	std::string description() override
	{
		return WindowDecorator::description() + ", including vertical scrollbars";
	}

private:
	void drawVerticalScrollBar()
	{
		std::cout << "Drawing Vertical ScrollBar" << std::endl;
	}
};

class HorizontalScrollBarDecorator : public WindowDecorator
{
public:
	HorizontalScrollBarDecorator(Widget* windowToDecorate)
		: WindowDecorator(windowToDecorate)
	{
	}

	void draw() override
	{
		WindowDecorator::draw();
		drawHorizontalScrollBar();
	}

	std::string description() override
	{
		return WindowDecorator::description() + ", including horizontal scrollbars";
	}

private:
	void drawHorizontalScrollBar()
	{
		std::cout << "Drawing Horizontal ScrollBar" << std::endl;
	}
};

int main()
{
	std::cout << ":: Decorator ::\n" << std::endl;

	Window* decoratedWindow = new HorizontalScrollBarDecorator(new VerticalScrollBarDecorator(new Window()));

	std::cout << decoratedWindow->description() << std::endl;

	system("PAUSE");

	delete decoratedWindow;

	return 0;
}

File: structural/PHP/Decorator.php
<?php

interface Widget {

    public function draw();

    public function getDescription();

}

class Window implements Widget {

    // override
    public function draw() {
        echo "Drawing the Window\n";
    }

    // override
    public function getDescription() {
        return 'Window';
    }

}

class WindowDecorator extends Window {

    private $windowToBeDecorated;

    public function __construct($windowToBeDecorated) {
        $this->windowToBeDecorated = $windowToBeDecorated;
    }

    // override
    public function draw() {
        $this->windowToBeDecorated->draw(); // Delegation
    }

    // override
    public function getDescription() {
        return $this->windowToBeDecorated->getDescription(); // Delegation
    }

}

class VerticalScrollBarDecorator extends WindowDecorator {

    public function __construct($windowToBeDecorated) {
        parent::__construct($windowToBeDecorated);
    }

    // override
    public function draw() {
        parent::draw();
        $this->drawVerticalScrollBar();
    }

    // override
    public function getDescription() {
        return parent::getDescription() . ', including vertical scrollbar';
    }

}

class HorizontalScrollBarDecorator extends WindowDecorator {

    public function __construct($windowToBeDecorated) {
        parent::__construct($windowToBeDecorated);
    }

    // override
    public function draw() {
        parent::draw();
        $this->drawHorizontalScrollBar();
    }

    // override
    public function getDescription() {
        return parent::getDescription() . ', including horizontal scrollbars';
    }

    private function drawHorizontalScrollBar() {
        echo "Drawing Horizontal ScrollBar\n";
    }

}

echo ":: Decorator ::\n";

$decoratedWindow = new HorizontalScrollBarDecorator(new VerticalScrollBarDecorator(new Window()));

echo $decoratedWindow->getDescription() . "\n";

?>